compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -324,7 +334,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
         let call = if self
             .get_resolution_id(delegation.id, span)
-            .and_then(|def_id| Ok(self.has_self(def_id, span)))
+            .and_then(|def_id| Ok(self.has_self(def_id, span) && self.is_assoc_fn(def_id)))
```

Comment:
It's better to fix this inside `fn has_self` itself, by checking `def_kind`, it will work on local `DefId`s too.

Reply:
Done. I have also renamed "has_self" to "is_method". Terminology for "method" came from [this](https://github.com/rust-lang/rfcs/pull/3530#discussion_r1411176926) conversation.